### PR TITLE
dev: Develop Payment Service and Wallet logic.

### PR DESCRIPTION
**Pull request by kai890707**
**Description**

1.  Implement Payment service logic.
2.  Implement Wallet logic.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] Conforms to style guide

### CODE REVIEW

File: develop/app/Controllers/v2/PaymentController.php [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    /**
+     * [GET] /api/v1/payment/{paymentKey}
+     * Get payment data by payment key.
+     *
+     * @param int $paymentKey
+     * @return void
+     */
+    public function show($paymentKey = null)
+    {
+        if ($paymentKey == null) {
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:84]: Add the type judge is better design.

File: develop/app/Controllers/v2/PaymentController.php [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        $total = $amount * $price;
+
+        $paymentModel  = new PaymentModel();
+        $walletModel   = new WalletModel();
+
+        $paymentEntity = $paymentModel->where("u_key", $u_key)
+                                      ->where("o_key", $o_key)
+                                      ->first();
+        if (!is_null($paymentEntity)) {
+            return $this->fail("This payment is exist , please check again.", 403);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:136]: Ditto.

File: develop/app/Controllers/v2/PaymentController.php [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        }
+
+        $userWallet = $walletModel->where('u_key', $u_key)
+                                  ->first();
+        if (is_null($userWallet)) {
+            return $this->fail("This user not exist", 400);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:142]: This user isn't exist.

File: develop/app/Controllers/v2/PaymentController.php [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+        $userBalance = $userWallet->balance;
+
+        if ($userBalance < $total) {
+            return $this->fail("Insufficient balance", 400);
+        }
+
+        $paymentCreatedIDOrNull = $paymentModel->createPaymentTransaction($u_key, $o_key, $total, $userBalance, $status);
+        if (is_null($paymentCreatedIDOrNull)) {
+            return $this->fail("Payment fail by payment data error", 400);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:153]: Payment created failed.

File: develop/app/Controllers/v2/PaymentController.php [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        }
+
+        $total  = $data["total"];
+        $status = $data["status"] ?? "paymentUpdate";
+
+        $paymentModel  = new PaymentModel();
+
+        $paymentEntity = PaymentModel::getPayment($paymentKey, $this->u_key);
+        if (is_null($paymentEntity)) {
+            return $this->fail("This payment is exist , please check again.", 404);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:189]: Ditto

File: develop/app/Models/v2/PaymentModel.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    protected $afterDelete    = [];
+
+    /**
+     * get someone payment data
+     *
+     * @param integer $paymentKey
+     * @param integer $userKey
+     * @return PaymentEntity|null
+     */
+    public static function getPayment(int $paymentKey, int $userKey): ?PaymentEntity
```
Comment (ping-yee) [inline develop/app/Models/v2/PaymentModel.php:51]: Ditto.

**Review by ping-yee — CHANGES_REQUESTED** [review]
Something need to be changed.

File: develop/app/Controllers/v2/PaymentController.php (hunk 1 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:84]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+     * @param int $paymentKey
+     * @return void
+     */
+    public function show($paymentKey = null)
+    {
+        if ($paymentKey == null) {
+            return $this->fail("The payment key is required.", 404);
+        }
+
+        $paymentEntity = PaymentModel::getPayment($paymentKey, $this->u_key);
```
[the remark this answers is not in the document]
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:88]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        if (is_null($paymentEntity)) {
+            return $this->fail("This payment is exist , please check again.", 404);
```
[the remark this answers is not in the document]
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:90]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 2 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:136]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 3 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:142]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 4 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:153]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 5 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php:189]: This change is done.

File: develop/app/Models/v2/PaymentModel.php (hunk 6 above)
Comment (kai890707) [reply develop/app/Models/v2/PaymentModel.php:51]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+class PaymentController extends BaseController
+{
+    use ResponseTrait;
+
+    private $u_key;
+
+    public function __construct()
+    {
+        $this->u_key     = UserService::getUserKey();
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:21]: Indentation.

File: develop/app/Controllers/v2/PaymentController.php [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+    /**
+     * [GET] /api/v1/payment/{paymentKey}
+     * Get payment data by payment key.
+     *
+     * @param int $paymentKey
+     * @return void
+     */
+    public function show($paymentKey = null)
+    {
+        if (is_null($paymentKey)) {
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:84]: My mean is using `===` to judge the type but this function invoke is fine.

File: develop/app/Controllers/v2/PaymentController.php [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+            return $this->fail("The payment key is required.", 404);
+        }
+
+        $paymentModel = new PaymentModel();
+
+        $paymentEntity = $paymentModel->where("u_key",$this->u_key)
+                                      ->find($paymentKey);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:91]: My mean before is the `getPayment` shouldn't be implemented in model, I think this method should be implemented in this controller.

File: develop/app/Controllers/v2/PaymentController.php [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+
+        $paymentEntity = $paymentModel->where("u_key", $u_key)
+                                      ->where("o_key", $o_key)
+                                      ->first();
+        if (!is_null($paymentEntity)) {
+            return $this->fail("This payment information is not exist or cannot found.", 403);
+        }
+
+        $userWallet = $walletModel->where('u_key', $u_key)
+                                  ->first();
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:143]: Does the logic of get user information has been implement at `WalletModel`?

File: develop/app/Controllers/v2/PaymentController.php [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+            return $this->fail("Incoming data not true", 400);
+        }
+
+        $total  = $data["total"];
+        $status = $data["status"] ?? "paymentUpdate";
+
+        $paymentModel  = new PaymentModel();
+
+        $paymentEntity = $paymentModel->where("u_key", $this->u_key)
+                                      ->find($paymentKey);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:191]: Same. This logic should be extracted, the reusability is too high.

**Review by ping-yee — CHANGES_REQUESTED** [review]
It's still something need to be changed.

File: develop/app/Controllers/v2/PaymentController.php (hunk 9 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 10 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 11 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 12 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 13 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+     */
+    public function delete($paymentKey = null)
+    {
+        if (is_null($paymentKey)) {
+            return $this->fail("The payment key is required.", 404);
+        }
+
+        $paymentModel  = new PaymentModel();
+
+        $paymentEntity = $paymentModel->where("u_key", $this->u_key)
```
[the remark this answers is not in the document]
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+     */
+    public function show($paymentKey = null)
+    {
+        if ($paymentKey === null) {
+            return $this->fail("The payment key is required.", 404);
+        }
+
+        $paymentModel = new PaymentModel();
+
+        $paymentEntity = $this->getPaymentEntityByPaymentKey($paymentKey);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:90]: This method name is too long, may use `getPaymentEntity`.

File: develop/app/Controllers/v2/PaymentController.php [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+        $total = $amount * $price;
+
+        $paymentModel  = new PaymentModel();
+        $walletModel   = new WalletModel();
+
+        $paymentEntity = $paymentModel->where("u_key", $u_key)
+                                      ->where("o_key", $o_key)
+                                      ->first();
+        if (!is_null($paymentEntity)) {
+            return $this->fail("This payment information is not exist or cannot found.", 403);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:138]: is not information` and `cannot found` mean the same thing, doesn't it?

File: develop/app/Controllers/v2/PaymentController.php [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+        if (is_null($data["total"]) || is_null($paymentKey)) {
+            return $this->fail("Incoming data not true", 400);
+        }
+
+        $total  = $data["total"];
+        $status = $data["status"] ?? "paymentUpdate";
+
+        $paymentModel  = new PaymentModel();
+
+        $paymentEntity = $this->getPaymentEntityByPaymentKey($paymentKey);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:188]: Same.

File: develop/app/Controllers/v2/PaymentController.php [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+     */
+    public function delete($paymentKey = null)
+    {
+        if (is_null($paymentKey)) {
+            return $this->fail("The payment key is required.", 404);
+        }
+
+        $paymentModel  = new PaymentModel();
+
+        $paymentEntity = $this->getPaymentEntityByPaymentKey($paymentKey);
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:223]: Same.

File: develop/app/Controllers/v2/PaymentController.php [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+        }
+    }
+
+    /**
+     * Get payment entity by payment key
+     *
+     * @param int $payment_key
+     * @return PaymentEntity|null
+     */
+    public function getPaymentEntityByPaymentKey($payment_key = null): ?PaymentEntity
```
Comment (ping-yee) [inline develop/app/Controllers/v2/PaymentController.php:253]: Re-name this method and limit the type of incoming data like: `int $payment_key`.

**Review by ping-yee — CHANGES_REQUESTED** [review]
Something need to be changed.

File: develop/app/Controllers/v2/PaymentController.php (hunk 15 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 16 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 17 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 18 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

File: develop/app/Controllers/v2/PaymentController.php (hunk 19 above)
Comment (kai890707) [reply develop/app/Controllers/v2/PaymentController.php]: This change is done.

**Review by ping-yee — APPROVED** [review]
LGTM!

**Pull request merged**